src/librustc_typeck/collect.rs

```diff
@@ -1511,9 +1511,29 @@ pub fn checked_type_of(tcx: TyCtxt<'_>, def_id: DefId, fail: bool) -> Option<Ty<
         }
 
         Node::GenericParam(param) => match &param.kind {
-            hir::GenericParamKind::Type { default: Some(ref ty), .. } |
-            hir::GenericParamKind::Const { ref ty, .. } => {
-                icx.to_ty(ty)
+            hir::GenericParamKind::Type { default: Some(ref ty), .. } => icx.to_ty(ty),
+            hir::GenericParamKind::Const { ty: ref hir_ty, .. } => {
```

Comment:
@varkor do we have a type whitelist somewhere? I remember there being a thing where we only allow specific types for const generics

Reply:
It hasn't been implemented yet: https://github.com/rust-lang/rust/issues/60286.